File: src/main/java/Transcript.java
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;

import java.io.FileWriter;
import java.io.IOException;
import java.io.PrintWriter;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;

public class Transcript extends Student {

    private List<Courses> completedCourses;
    private List<Courses> failedCourses;
    private Double gpa;
    private int completedCredits;
    private String advisorName;
    private List<String> studentSelectedCourses;
    private List<String> completedCourseStrings;
    private List<String> failedCoursesStrings;

    public Transcript() {
    }
    public Transcript(List<Courses> completedCourses, List<Courses> failedCourses, Double gpa, int completedCredits, List<String> studentSelectedCourses, String advisorName, Student student){
        this.completedCourses = completedCourses;
        this.failedCourses = failedCourses;
        this.gpa = gpa;
        this.completedCredits = completedCredits;
        this.advisorName = advisorName;
        this.studentSelectedCourses = studentSelectedCourses;

    }

    public List<Courses> getCompletedCourses() {
        return completedCourses;
    }
    public List<Courses> getFailedCourses() {
        return failedCourses;
    }
    /*public void printTranscriptAll(Student[] students2){
        for (Student s: students2) {
            System.out.println("ID: "+s.getStudentId() + "\nName: " + s.getfName() + " " +  s.getlName() + "\nCourses: " + s.getCompletedCourses() + "\nSelected Courses:" + s.getCurrentSelectedCourses() + "\nGPA: " + s.getGPA() + "\nTotalCredits: " + s.getTotalCredit() + "\n-------------------------------------------------------------------------------------------------------------------------------------------------");
        }
    }*/
    public void seperateFailedCourses(){
        failedCoursesStrings = new ArrayList<>();
        for (Courses completedCourse : this.completedCourses){
            if (completedCourse.getCourseGrade().equals("FF")){
                failedCoursesStrings.add(completedCourse.getName());
                failedCoursesStrings.add(completedCourse.getCourseGrade());
            }

        }
    }
    public void printTranscriptSpecificStudent(Student student){
        student.getTranscript().transformSpecificStudentTranscriptElementsToList(student);
        System.out.println("------------------------------------------------------------------------------------------------");
        System.out.println("ID: " + student.getStudentId() + "\nFullName: " + student.getfName() + " " + student.getlName() + "\nAdvisor: " + this.advisorName + "\nCourses Taken:" + this.completedCourseStrings + "\nGPA: " + this.gpa + "\nTotal Credits: " + this.completedCredits + "\nAvailable Courses:" + student.getAvailableCourses()  + "\nAdvisor Approved Courses: " + this.studentSelectedCourses + "\nFailed Courses: " + this.failedCoursesStrings );
    }
    public void transformTranscriptElementsToList(){
        this.completedCourseStrings = new ArrayList<>();
        for (Courses completedCourses : this.completedCourses){
            completedCourseStrings.add(completedCourses.getName());
            completedCourseStrings.add(completedCourses.getCourseGrade());
        }
    }
    public void transformSpecificStudentTranscriptElementsToList(Student student){
        this.completedCourseStrings = new ArrayList<>();
        for (Courses completedCourses : student.getCompletedCourses()){
            completedCourseStrings.add(completedCourses.getName() + " " + completedCourses.getCourseGrade() + " Given: " + completedCourses.getGivenSemester());
        }
        seperateFailedCourses();
    }

    public void generateTranscriptJson(Student[] students) throws IllegalAccessException {
        JSONArray arr=new JSONArray();
        for (Student std:students){
            JSONObject jsonObj=new JSONObject();
            for (Field f:std.getClass().getDeclaredFields()){
                f.setAccessible(true);
                if(!f.getName().equals("failedCourses") && !f.getName().equals("availableCourses"))
                    jsonObj.put(f.getName(),f.get(std));
            }
            arr.add(jsonObj);
        }

        try (PrintWriter out = new PrintWriter(new FileWriter("./StudentsOutput.json"))){
            out.write(arr.toString());
        } catch (Exception e){
            e.printStackTrace();
        }
    }
}

File: transcript.py
import json


class Transcript:
    def __init__(self, completedCourses, failedCourses, gpa, completedCredits, studentSelectedCourses, student):
        self.completedCourses = completedCourses
        self.failedCourses = failedCourses
        self.gpa = gpa
        self.completedCredits = completedCredits
        self.studentSelectedCourses = studentSelectedCourses
        self.student = student
        self.failedCoursesStrings = []
        self.completedCourseStrings = []

    def separate_failed_courses(self):
        for completedCourse in self.completedCourses:
            if completedCourse.get_course_grade == "FF":
                self.failedCoursesStrings.append(completedCourse.get_course_name())
                self.failedCoursesStrings.append(completedCourse.get_course_grade())

    def print_transcript_specific_student(self, student):
        student.transcript.transform_specific_student_transcript_elements_to_list(student)
        print("------------------------------------------------------------------------------------------------")
        print(str("ID: {}\nFullName: {}\nCourses Taken: {}\nGPA: {}\nTotal Credits: {}\nAvailable Courses: {}\nAdvisor Approved Courses: {}\nFailed Courses: {}".format(student.get_student_id(), student.get_full_name(), self.completedCourseStrings, self.gpa, self.completedCredits, student.get_available_courses(), self.studentSelectedCourses, self.failedCoursesStrings)))

    def transform_transcript_elements_to_list(self):
        self.completedCourseStrings = []
        for completedCourses in self.completedCourses:
            self.completedCourseStrings.append(completedCourses.get_course_name())
            self.completedCourseStrings.append(completedCourses.get_course_grade())

    def transform_specific_student_transcript_elements_to_list(self, student):
        for completedCourses in student.get_completed_courses():
            self.completedCourseStrings.append(str("{} {} Given: {}".format(completedCourses.get_course_name(), completedCourses.get_course_grade(), completedCourses.get_given_semester())))

    def save_transcript_to_json(self):
        data = {}
        data['completedCourses'] = []
        for completedCourse in self.completedCourses:
            data['completedCourses'].append({
                'courseName': completedCourse.get_course_name(),
                'courseGrade': completedCourse.get_course_grade(),
                'givenSemester': completedCourse.get_given_semester()
            })
        data['failedCourses'] = []
        for failedCourse in self.failedCourses:
            data['failedCourses'].append({
                'courseName': failedCourse.get_course_name(),
                'courseGrade': failedCourse.get_course_grade(),
                'givenSemester': failedCourse.get_given_semester()
            })
        data['gpa'] = self.gpa
        data['completedCredits'] = self.completedCredits
        data['studentSelectedCourses'] = self.studentSelectedCourses
        data['student'] = {
            'studentId': self.student.student_id,
            'fName': self.student.f_name,
            'lName': self.student.l_name,
            'advisor': self.student.get_advisor_name()
        }

        with open('transcript.json', 'w') as outfile:
            json.dump(data, outfile)
